### ISSUE TITLE
sql: internal query to read role memberships encounters GC threshold error [breaks user logins, `SHOW GRANTS`, and more]

### ISSUE THREAD

**Issue by THardy98** [cockroachdb/cockroach#102385]
Related to: #102204

**Describe the problem**

Trying to get tables' grants on the test cluster using `SHOW GRANTS ON <table>` results in a GC threshold error:
```
read-members: batch timestamp 1681781325.979933937,0 must be after replica GC threshold 1681921008.470061491,0
```

**To Reproduce**

Run any `SHOW GRANTS ON <table name>` query on the test cluster.

Jira issue: CRDB-27419

**Comment by rafiss** [comment]
This is quite bad. The underlying problem is that the internal operation to read role memberships from `system.role_members` is not working. The impact of that is much larger. Specifically, this means no new user can login.
Right now `root` is the only user being used in testing right now, and its information is cached, so logging in as `root` does work. But if that cache gets invalidated, I think even `root` logins could fail too.

I made a new user and confirmed the problem:
```
❯ roachprod sql grinaker-231:1 -- --user=testuser_rafi
...
ERROR: internal error while retrieving user account memberships: read-members: batch timestamp 1681781325.979933937,0 must be after replica GC threshold 1681921008.470061491,0
SQLSTATE: 28000
Failed running "sql"
Connection to 34.139.211.171 closed.
```

**Comment by rafiss** [comment]
This is the internal query that fails: https://github.com/cockroachdb/cockroach/blob/f1a130b93c47ddd9a11408d23f48cb6ee96eb38e/pkg/sql/information_schema.go#L2765-L2771

(it's called from [here](https://github.com/cockroachdb/cockroach/blob/f1a130b93c47ddd9a11408d23f48cb6ee96eb38e/pkg/sql/authorization.go#L496).)

But running the query manually as `root` does work:
```
❯ roachprod sql grinaker-231:1
#
# Welcome to the CockroachDB SQL shell.
# All statements must be terminated by a semicolon.
# To exit, type: \q.
#
# Server version: CockroachDB CCL v23.1.0-beta.4-dev-ec0125cca00d0f924d2522926bb5f3c328c4e372 (x86_64-pc-linux-gnu, built 2023/04/25 16:52:13, go1.19.4) (same version as client)
# Cluster ID: 32c936d5-7732-423f-be38-196b7f0afcb6
# Organization: Cockroach Labs - Production Testing
#
# Enter \? for a brief introduction.
#
root@localhost:26257/defaultdb> SELECT "role", "member", "isAdmin" FROM system.role_members;
  role  | member | isAdmin
--------+--------+----------
  admin | root   |    t
(1 row)
```

So maybe something is wrong with the internal executor.

**Comment by rafiss** [comment]
Here are logs I grabbed after hitting the error:
```
W230427 03:58:57.585700 1181449245 sql/user.go:167 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13815  user membership lookup for ‹"testuser_rafi"› failed: ‹read-members›: batch timestamp 16
81781325.979933937,0 must be after replica GC threshold 1681921008.470061491,0
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816  user retrieval failed for user=‹"testuser_rafi"›: internal error while retrievin
g user account memberships: ‹read-members›: batch timestamp 1681781325.979933937,0 must be after replica GC threshold 1681921008.470061491,0
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +(1) attached stack trace
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  -- stack trace:
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  | github.com/cockroachdb/cockroach/pkg/sql.GetUserSessionInitInfo
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |      github.com/cockroachdb/cockroach/pkg/sql/user.go:168
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).handleAuthentication
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |      github.com/cockroachdb/cockroach/pkg/sql/pgwire/auth.go:149
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
.func1
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |      github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:700
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  | runtime.goexit
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |      GOROOT/src/runtime/asm_amd64.s:1594
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +Wraps: (2) internal error while retrieving user account memberships
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +Wraps: (3) ‹read-members›: batch timestamp 1681781325.979933937,0 must be after
replica GC threshold 1681921008.470061491,0
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  | -- cause hidden behind barrier
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  | ‹read-members›: batch timestamp 1681781325.979933937,0 must be after replica
 GC threshold 1681921008.470061491,0
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  | (1) secondary error attachment
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   | ‹read-members›: batch timestamp 1681781325.979933937,0 must be after rep
lica GC threshold 1681921008.470061491,0
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   | (1) attached stack trace
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   -- stack trace:
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInt
ernal.func1.1
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/internal.go:971
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.(*rowsIterator).Next.func1
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/internal.go:418
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.(*rowsIterator).Next
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/internal.go:469
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.forEachRoleMembership
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:
2777
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.resolveMemberOfWithAdminOpt
ion
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/authorization.go:657
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.MemberOfWithAdminOption.fun
c3.1
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/authorization.go:590
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn.func1
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/internal.go:1604
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/internal.go:1682
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/kv/db.go:965
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/kv/txn.go:928
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/kv.runTxn
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/kv/db.go:964
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/kv/db.go:927
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/kv/db.go:902
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/internal.go:1670
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/internal.go:1605
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/sql.MemberOfWithAdminOption.func3
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/sql/authorization.go:585
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall.func1
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:387
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTask
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:305
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:386
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   | runtime.goexit
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   |   |      GOROOT/src/runtime/asm_amd64.s:1594
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   | Wraps: (2) ‹read-members›
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   | Wraps: (3)
W230427 03:58:57.585819 1181449245 sql/pgwire/auth.go:156 ⋮ [T1,n1,client=127.0.0.1:34036,user=‹testuser_rafi›] 13816 +  |   | Wraps: (4) batch timestamp 1681781325.979933937,0 must be after replica GC threshold 1681921008.470061491,0
```

The issue could be with the internal executor or `singleflight` (or something else).

**Comment by tbg** [comment]
More discussion that is closer to resolution [here on Slack](https://cockroachlabs.slack.com/archives/C04QZ9Z1QR1/p1682568385438839)

**Comment by tbg** [comment]
Very likely root cause here

https://cockroachlabs.slack.com/archives/C04QZ9Z1QR1/p1682583086276619?thread_ts=1682568385.438839&cid=C04QZ9Z1QR1

TLDR SQL is issuing reads at the modtime of a table desc, but KV may not support reads at that timestamp any more. The read timestamp that gets rejected matches exactly the modtime of the `role_members` table.

**Comment by rafiss** [comment]
Thanks @tbg; https://github.com/cockroachdb/cockroach/pull/98617 is definitely the culprit. I've merged a fix in the linked PRs, but also for posterity I'm leaving a note below on how to unbreak the cluster without code changes.

---

I ran this descriptor surgery to get out of the broken state. This query bumps the table version of `system.role_members`, which will cause it to get a new modification time, which will cause the operation to read role memberships to use a newer non-expired timestamp.

Normally a `GRANT role TO user` command would cause the `system.role_members` mod time to be changed too, but the bug here also breaks `GRANT` statements.

```
WITH
	role_members_desc
		AS (
			SELECT
				d.id,
				crdb_internal.pb_to_json(
					'cockroach.sql.sqlbase.Descriptor',
					d.descriptor
				)
					AS desc
			FROM
				system.descriptor AS d
				JOIN system.namespace AS n ON n.id = d.id
			WHERE
				n.name = 'role_members' AND n."parentID" = 1
		)
SELECT
	crdb_internal.unsafe_upsert_descriptor(
		role_members_desc.id,
		crdb_internal.json_to_pb(
			'cockroach.sql.sqlbase.Descriptor',
			json_set(
				role_members_desc.desc,
				'{table, version}',
				(
					btrim(
						role_members_desc.desc['table']['version']::STRING,
						'"'
					)::INT8
					+ 1
				)::STRING::JSONB
			)
		)
	)
FROM
	role_members_desc;
```
